nmigen/hdl/rec.py

```diff
@@ -166,11 +175,23 @@ def __getitem__(self, item):
                 if field_name in item
             })
         else:
-            return super().__getitem__(item)
+            try:
+                return Value.__getitem__(self, item)
+            except KeyError:
+                if self.name is None:
+                    reference = "Unnamed record"
+                else:
+                    reference = "Record '{}'".format(self.name)
+                raise AttributeError("{} does not have a field '{}'. Did you mean one of: {}?"
+                                     .format(reference, item, ", ".join(self.fields))) from None
 
-    def lower(self):
+    @ValueCastable.lowermethod
+    def as_value(self):
         return Cat(self.fields.values())
 
+    def __len__(self):
+        return len(Cat(self.fields.values()))
```

Comment:
Doing `len(self.as_value())` here would result in less garbage because of memoization (nMigen calls len() a lot so it might actually matter).